### PR TITLE
Feature/api/idea database

### LINKED ISSUE #537: Funcionalidades Banco de Idéias
## Feature Description

O banco de ideias é uma funcionalidade disponível para todos os usuários registrarem suas demandas de pesquisas e tecnologias. Vai ser a oportunidade para aqueles que não encontraram a tecnologia que buscam e sugerir que os pesquisadores desenvolvam de acordo com a demanda da sociedade.

O banco de ideias será composto por:

**1 - Lista corrida de oportunidades cadastradas pelos usuários**
- Uma pequena descrição abaixo do título da página:
_Não encontrou o que desejava?
Aqui você pode sugerir novas ideias de desenvolvimento de pesquisa e tecnologia de acordo com sua necessidade e para resolver problemas comuns do nosso povo. Veja abaixo as ideias já sugeridas pelos nossos usuários e registre a sua também._

- Lista simples composta pelo seguinte conteúdo
  - Título da ideia (texto)
  - Detalhe da ideia (textarea)
  - Palavras-chave
  - Usuário que sugeriu a ideia
  - Data do registro
- Incluir opções para filtrar a lista por texto do título e descrição e por palavra-chave
- Essa lista será acessar por meio do menu principal da Plataforma (estamos definindo esse menu ainda)

**2 - Cadastrar uma ideia**
- Somente usuários cadastrados podem registrar uma ideia
- Quando clicar em REGISTRAR UMA IDEIA aparecer uma tela com os seguintes campos:
- TÍTULO DA IDEIA
- DESCRIÇÃO DA IDEIA (textarea): 
- PALAVRAS-CHAVE (o mesmo do cadastro de tecnologias)
- botão ENVIAR IDEIA
 
**3 - Lista de Ideias para os usuários pesquisadores**
- Na área do pesquisador incluir um item no menu chamado BANCO DE IDEIAS
- Apresentar os seguintes campos na tabela:
  - Título
  - Usuário
  - Data
  - Palavras-chave
- Incluir opções para filtrar a lista por texto do título e descrição e por palavra-chave
- Incluir um ícone de detalhes que apresenta um modal com a descrição da ideia

---------------

## Acceptance criteria

* 
- Somente usuários logados e cadastrados poderão registrar novas ideias.

## Implementation Brief

*

### PR DESCRIPTION
**Pull request by alexandreadames**
## Summary

Resolves #537

### Rotas

| Description | Method | Route | Body/Query | Auth |
| --- | --- | --- | --- | --- |
Lists all Ideas. |GET| ideas | Query Params for filtering: title, description, keywords | No | 
|Shows an idea. |GET| idea/:id | N/A | No |
|Any user with complete profile can creates an idea |POST| ideas | title (string),description (string),  keywords (term array id or slug) | Yes |
|Idea owner user can update it |PUT| idea/:id | title (string),description (string),  keywords (term array id or slug) | Yes Permission: UPDATE_IDEA|
|Idea owner user can delete it |DELETE| idea/:id | N/A | Yes Permission: DELETE_IDEA|

## QA Steps

Run `test/functional/idea.spec.js`

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code passes the linting.
- [x] My code has proper inline documentation.
- [x] I have manually tested this PR.
- [x] Documentation with ApiDOC

### CODE REVIEW

**Review by NaylsonFerreira — APPROVED** [review]
Padrão... Padrão... 👌

File: packages/api/database/migrations/1606773532660_idea_schema.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+	up() {
+		this.create('ideas', (table) => {
+			table.increments();
+			table.string('title').notNullable();
+			table.text('description').notNullable();
+			table
+				.integer('user_id')
+				.unsigned()
+				.references('id')
+				.inTable('users');
```
Comment (mateus4k) [inline packages/api/database/migrations/1606773532660_idea_schema.js:14]: Faltou o onUpdate e onDelete

Comment (mateus4k) [inline packages/api/database/migrations/1606773532660_idea_schema.js:14]: Faltou o onUpdate e onDelete

**Review by mateus4k — CHANGES_REQUESTED** [review]
Good job, man! 👊

**Review by mateus4k — APPROVED** [review]

**Comment by mateus4k** [comment]
Se lembre de aplicar a mudança que comenteia

File: packages/api/app/Validators/StoreIdea.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+const BaseValidator = use('App/Validators/BaseValidator');
+
+class StoreIdea extends BaseValidator {
+	get rules() {
+		return {
+			title: 'string|required',
+			description: 'string|required',
+			keywords: 'array|required',
```
[the remark this answers is not in the document]
Comment (alexandreadames) [reply packages/api/app/Validators/StoreIdea.js:8]: @NaylsonFerreira nesse caso ele aceita number(id) ou string(term slug), você sabe como especificar dois tipos ao mesmo tempo? Eu testei fazer assim `keywords.*: string|number`, mas nessse caso ele dá problema que não seria um ou outro e sim um e outro.

**Review by mateus4k — DISMISSED** [review]

**Comment by NaylsonFerreira** [comment]
Falta ver os testes que falharam

File: packages/api/app/Validators/StoreIdea.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (NaylsonFerreira) [reply packages/api/app/Validators/StoreIdea.js:8]: Não sei como fazer pra checar os dois.
Imaginei que seria só inteiros

**Review by NaylsonFerreira — APPROVED** [review]

**Review by mateus4k — APPROVED** [review]

**Pull request merged**